### PR TITLE
A `SubstitutionPart` is not considered a deletion if it replaces nothing with nothing

### LINKED ISSUE #101689: Improve and make missing `dyn` help suggestion more uniform
Given the following code: [link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=87f85ad7ab15806e099c71a26587c1b5)

```rust
trait A { }

impl A {

}

fn main() {

}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0782]](https://doc.rust-lang.org/nightly/error-index.html#E0782): trait objects must include the `dyn` keyword
 --> src/main.rs:3:6
  |
3 | impl A {
  |      ^
  |
help: add `dyn` keyword before this trait
  |
3 - impl A {
3 + impl dyn A {
  |

For more information about this error, try `rustc --explain E0782`.
error: could not compile `playground` due to previous error
```

Ideally the output should look like:

```
Compiling playground v0.0.1 (/playground)
error[[E0782]](https://doc.rust-lang.org/nightly/error-index.html#E0782): trait objects must include the `dyn` keyword
 --> src/main.rs:3:6
  |
3 | impl A {
  |      ^
  |
help: add `dyn` keyword before trait `A`
  |
3 | impl dyn A {
  |      +++

For more information about this error, try `rustc --explain E0782`.
error: could not compile `playground` due to previous error
```
Above changes: use a more common `+` convention for the help suggestion and make the target trait more explicit by mentioning it.
<hr>

Since most of the other help suggestion that requires adding something uses the `+` convention, I think we should also change the above suggestion to use that convention to make things more uniform. On top of that the new proposed output is less verbose as it reduces 1 line of output.
<hr>

Note: If the old "diff" convention for the help suggestion is intentional, please let me know the reason and I will gladly close this issue. Thanks.

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #101689

### CODE REVIEW

File: src/test/ui/rust-2021/reserved-prefixes-migration.stderr [hunk 1]
```diff
@@ -41,8 +39,7 @@ LL |     m3!(hey#123);
    = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/reserving-syntax.html>
 help: insert whitespace here to avoid this being parsed as a prefix in Rust 2021
    |
-LL -     m3!(hey#123);
-LL +     m3!(hey #123);
+LL |     m3!(hey #123);
```
Comment (compiler-errors) [inline src/test/ui/rust-2021/reserved-prefixes-migration.stderr:42]: hm, this is a bug actually

File: src/test/ui/rust-2021/reserved-prefixes.stderr [hunk 2]
```diff
@@ -85,9 +79,8 @@ LL |     demo4!(foo## bar);
    = note: prefixed identifiers and literals are reserved since Rust 2021
 help: consider inserting whitespace here
    |
-LL -     demo4!(foo## bar);
-LL +     demo4!(foo ## bar);
-   |
+LL |     demo4!(foo ## bar);
```
Comment (compiler-errors) [inline src/test/ui/rust-2021/reserved-prefixes.stderr:82]: This change is much better <3

**Comment by lcnr** [comment]
r? compiler

**Review by davidtwco — APPROVED** [review]

**Review by davidtwco — APPROVED** [review]

**Pull request merged**